### PR TITLE
Fix list BParser

### LINKED ISSUE #2: list parser should not require name parameter
The list parser currently is based on the dictionary parser and requires a `String` name parameter. This prevents parsing lists of lists, which are used in, for example, BEP-12: http://bittorrent.org/beps/bep_0012.html

### PR DESCRIPTION
**Pull request by matthewleon**
Closes #2. Note that this changes the API for BParser. A more conservative approach might be to deprecate the old "list" parser in favor of "list_" or something like it.

Tests included.

### CODE REVIEW

File: src/Data/BEncode/Parser.hs [hunk 1]
```diff
@@ -82,12 +82,11 @@ dict name = BParser $ \b -> case b of
...
-    = dict name >>= \lst ->
-      BParser $ \b -> case lst of
-                      BList bs -> foldr (cat . runB p) (Ok [] b) bs
-                      _ -> Error $ "Not a list: " ++ name
+list :: BParser a -> BParser [a]
+list p
+    = BParser $ \lst -> case lst of
+          BList (bs) -> foldr (cat . runB p) (Ok [] (BList [])) bs
+          _ -> Error $ "Not a list: " ++ (show lst)
     where cat (Ok v _) (Ok vs b) = Ok (v:vs) b
```
Comment (creichert) [inline src/Data/BEncode/Parser.hs:90]: Could this potentially blow up if for some reason the input was unexpectedly large or malformed?

Comment (matthewleon) [inline src/Data/BEncode/Parser.hs:90]: Can you explain? Is it the `(Ok [] (BList []))` parameter to `foldr` you are looking at? I was thinking of just making that parameter `(Ok [] undefined)`, thinking that the second param to the `Ok` constructor never would get evaluated. If it's possible to make it blow up, I'm sure some simple input in ghci should find a case, no?

Comment (creichert) [inline src/Data/BEncode/Parser.hs:90]: Sorry, I was actually talking about the line 89 with the Error. Specifically, `show lst`.

Comment (matthewleon) [inline src/Data/BEncode/Parser.hs:90]: Yeah, that is potentially a nasty one. I'll add a fix.

Comment (matthewleon) [inline src/Data/BEncode/Parser.hs:90]: Now that I think of it, though, it seems no worse than the failure cases for `bint`, `bstring`, etc. So if this is to be fixed, it should probably be a separate issue, no?

Comment (creichert) [inline src/Data/BEncode/Parser.hs:90]: You're right, no fix necessary in this patch.

**Comment by creichert** [comment]
Fantastic, looks great. I'm guessing the list name was originally meant as a debug source.

If you're interested, here's a library that used the original list function (very bottom): https://hackage.haskell.org/package/torrent-10000.0.0/docs/src/Data-Torrent.html

**Comment by matthewleon** [comment]
Yes, I came to your library through that one. I will be updating it shortly so that it handles multiple trackers, which requires this change.

**Comment by creichert** [comment]
Very nice, I can coordinate a push to Hackage and a major version bump whenever is convenient.

**Pull request merged**